pkg/engine/task/task_apply.go

```diff
@@ -99,33 +137,87 @@ func apply(rr []runtime.Object, c client.Client, di discovery.DiscoveryInterface
 	return applied, nil
 }
 
-// patch calls update method on kubernetes client to make sure the current resource reflects what is on server
-//
-// an obvious optimization here would be to not patch when objects are the same, however that is not easy
-// kubernetes native objects might be a problem because we cannot just compare the spec as the spec might have extra fields
-// and those extra fields are set by some kubernetes component
-// because of that for now we just try to apply the patch every time
-// it mutates the object passed in to be consistent with the kubernetes client behavior
-func patch(newObj runtime.Object, c client.Client) error {
-	newObjJSON, _ := apijson.Marshal(newObj)
-	key, _ := client.ObjectKeyFromObject(newObj)
+func patch(updatedObj, currentObj runtime.Object, c client.Client) error {
+
+	// Serialize current configuration
+	current, err := json.Marshal(currentObj)
+	if err != nil {
+		return fmt.Errorf("failed to marshal current %v", err)
+	}
+
+	// Get previous configuration from currentObjs annotation
+	original, err := getOriginalConfiguration(currentObj)
+	if err != nil {
+		return fmt.Errorf("failed to get original configuration %v", err)
+	}
+
+	// Get new (modified) configuration
+	modified, err := getModifiedConfiguration(updatedObj, true, unstructured.UnstructuredJSONScheme)
+	if err != nil {
+		return fmt.Errorf("failed to get modified config %v", err)
+	}
+
+	// Create the actual patch
+	var patchData []byte
+	var patchType types.PatchType
+	if useJSONMerge(updatedObj) {
+		patchType = types.MergePatchType
+		patchData, err = jsonThreeWayMergePatch(original, modified, current)
+	} else {
+		patchType = types.StrategicMergePatchType
+		patchData, err = strategicThreeWayMergePatch(updatedObj, original, modified, current)
+	}
+
+	if err != nil {
+		// TODO: We could try to delete/create here, but that would be different behavior from before
+		return fmt.Errorf("failed to create patch: %v", err)
+	}
+
+	// Execute the patch
+	err = c.Patch(context.TODO(), updatedObj, client.ConstantPatch(patchType, patchData))
+	if err != nil {
+		return fmt.Errorf("failed to execute patch: %v", err)
+	}
+	return nil
+}
+
+func useJSONMerge(newObj runtime.Object) bool {
```

Comment:
nit:
```suggestion
func shouldMergeThreeWays(newObj runtime.Object) bool {
```
(or smth. similar)

Reply:
I've decided on `useSimpleThreeWayMerge`